[Docs] 4주차 발표자료 - #50

jang-namu opened the pull request:
## 📋 이슈 번호
close #50 
close #60 
close #64 
## 🛠 구현 사항
4주차 발표자료입니다.

+) Controller & Service 기본 로직 + swagger config
## 📚 기타

jang-namu committed:
[Docs] 4주차 발표자료

jang-namu committed:
feat: Task Service & Controllers basic feature

jang-namu committed:
feat: User Service & Controller basic feature

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/task/Task.java:
```diff
@@ -23,5 +29,7 @@ public class Task {
     private LocalDateTime deadline;
 
-    private Boolean is_completed;
+    @Setter
```
해당 필드만 set을 하더라도 setter 어노테이션은 지양하는게 좋아
이유는 스터디때 했으니 알거라고 생각해!

jang-namu replied:
그치 setter는 항상 문제가 되니까. 적어도 이름을 통해 행위를 적절하게 알리는 메서드로 대신하는게 좋겠네.
추가적으로 HTTP 중 PATCH 메서드를 어떻게 처리하는지도 찾아볼게.

Juser0 replied:
형이 set해야하는 행위에 기반한 네이밍으로 메소드를 구현하는 식이면 좋을 것 같아

jang-namu replied:
<img width="717" alt="image" src="https://github.com/inu-appcenter/server-study-15.5/assets/121238128/d7689da0-d262-45c9-ae77-23c0761c917d">

~~일단 위와같이 고쳐봤어. 괜찮을까? 메서드명이 좀 애매한거 같기도하고~~
다시 updateIsCompleted로 수정.
어차피 boolean 타입인 것도 이름에서 잘 들어난다고 생각되서

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/task/Task.java:
```diff
@@ -1,17 +1,23 @@
+@Getter
 @Entity
+@Builder
+@NoArgsConstructor
```
NoArgsConstructor 어노테이션에서 접근제한자를 지정하는게 좋은 이유가 뭐였지?

jang-namu replied:
그러네, 사실 Builder와 AllArgs..를 같이 사용할꺼라면
NoArgs..에는 accesslevel = protected를 지정하는게 객체 보호 관점에서 좋을 것 같다. 불완전한 객체 사용을 막으려면..

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/task/Task.java:
```diff
@@ -1,17 +1,23 @@
 @Table(name = "task_tb")
+@Getter
 @Entity
+@Builder
```
빌더 패턴을 사용하는건 좋은데 어노테이션에 의존하는 형태로 구현하는게 바람직한지 고민해봐

jang-namu replied:
<img width="897" alt="image" src="https://github.com/inu-appcenter/server-study-15.5/assets/121238128/99ebf0b8-b846-4fb4-848a-3f493707b143">

오키! 클래스 수준의 AllArgs..와 Builder를 삭제하고
생성자를 명시적으로 만들어서 Builder를 다는 걸로 수정헀어.

다만 '어노테이션에 의존하는 형태의 구현'이라는 게 아직 무슨 의미인지는 잘 이해를 못했다 ㅠ

Juser0 replied:
어노테이션에 의존하는 형태의 구현이라는게 클래스에 AllArgsConstructor + Builder로 빌더 패턴 / 생성자 패턴을 구현하는 걸 말하는 거였어!

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/task/TaskController.java:
```diff
@@ -0,0 +1,41 @@
+        taskService.delete(taskId);
+    }
+
+    @PatchMapping("/iscompleted/{taskId}")
```
HTTP Method를 통해 행위에 대한 내용을 파악할 수 있는데 앤드포인트에 행위를 명시하는게 좋을까?

jang-namu replied:
'iscompleted'라는 필드만을 변경하는 엔드포인트라 마땅한 이름이 생각이 안났는데, 좀 더 고민해볼게.

Juser0 replied:
완료여부만을 수정하는 메소드
전반적인 내용을 수정하는 메소드
이렇게 나뉘는 경우라면 엔드포인트를 그렇게 해주는 게 좋아보여 (구별용)
그런데 그렇지 않다면 그냥 Patch라는 메소드로 수정이라는 내용을 알거니깐 굳이 엔드포인트에 명시할 필요는 없어보여

jang-namu replied:
응, 우선 title이나 decription, deadline 등도 수정할 수 있게 하려고 하는데.
고민중인게, dto를 받아서 하나의 엔드포인트로 모두 수정할 수 있도록 할까.
아니면 각각의 필드마다 엔드포인트를 따로 만들어야할까. 고민이야.

Juser0 replied:
수정의 범위가 그러면
- 할 일에 대한 전반적인 수정 (title, description, deadline, isCompleted 혹은 완료여부 빼고 수정하도록) /api/tasks/{taskId}
- 완료 여부에 대한 수정 /api/tasks/completed/{taskId}
이렇게 하고 대강 이런 엔드포인트로 가면 되지 않을까?

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/task/TaskService.java:
```diff
@@ -0,0 +1,82 @@
+        List<TaskResponseDto> taskResponseDtos = new ArrayList<>();
+
+        List<Task> taskList = taskRepository.findAllByUser_UserId(userId);
+        for (Task task : taskList) {
```
ResponseDto 형식으로 리스트를 생성하는 경우에는 스트림을 사용해서 구현이 가능해

jang-namu replied:
ResponseDto에 Task를 매개변수로 받는 생성자를 정의하고 
Stream으로 바꿨어!

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/task/TaskService.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    public TaskResponseDto save(TaskDto taskDto) throws Exception {
+        Optional<User> selectedUser = userRepository.findById(taskDto.getUserId());
```
여기서 orElseThrow로 명시해주면 if문으로 present를 파악해줄 필요가 없어

jang-namu replied:
맞네, 수정했어! 내가봐도 if 문은 너무 올드했다. 고마워.

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/task/dto/TaskResponseDto.java:
```diff
@@ -0,0 +1,23 @@
+
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
```
AllArgsConstructor 어노테이션을 지양해야하는 이유는?

jang-namu replied:
맞네. AllArgs...는 파라미터의 순서가 보장되지 않아서 아래와 같은 경우.
title과 description이 뒤섞일 위험도 있겠다.

Juser0 replied:
그치 AllArgsConstructor 어노테이션은 아마 Builder 어노테이션을 위해 사용한 것 같은데,
우리가 배웠던 내용 생각해보면 형이 직접 매개변수가 있는 생성자 등을 정의하고 해당 생성자에 Builder 어노테이션을 붙여서 사용할 때는 같지만 어노테이션에 의존적인 형태는 피하는 게 좋지

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/task/dto/TaskResponseDto.java:
```diff
@@ -0,0 +1,23 @@
+import java.time.LocalDateTime;
+
+@Builder
+@NoArgsConstructor
```
매개변수가 없는 생성자를 응답 dto에서 사용하는 경우가 있어?

jang-namu replied:
없네. 아래 AllArgsConstructor와 함께 삭제하도록 할게.

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/user/UserController.java:
CRUD에서 Update가 구현되지 않은것 같아!

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/user/UserController.java:
```diff
@@ -0,0 +1,33 @@
+    @PostMapping()
+    public ResponseEntity<UserResponseDto> postUser(UserRequestDto userRequestDto) {
+        UserResponseDto userResponseDto = userService.saveUser(userRequestDto);
+        return ResponseEntity.status(HttpStatus.OK).body(userResponseDto);
```
POST method의 응답 코드로는 201 CREATED가 더 적절해!

jang-namu replied:
생각없이 따라친 결말인가... 수정하도록 할게!

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/user/dto/UserRequestDto.java:
requestdto에서 기본생성자가 없어도 정상적으로 동작했어?

jang-namu replied:
동작에 문제가 없는걸로 보면, 아무것도 지정하지 않아서 기본생성자가 자동 생성되는 것 같아!

NARUBROWN replied:
자동으로 생성되는게 맞아!

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/user/dto/UserRequestDto.java:
```diff
@@ -0,0 +1,16 @@
+package com.example.todo.user.dto;
+
```
여기 사용되지 않은 어노테이션에 대한 import 내용도 있어보이네

jang-namu replied:
[["Optimizing imports: ON"]]

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/task/TaskController.java:
task에서 수정할 부분이 isCompleted 뿐인지 고민해봐!

jang-namu replied:
전체 수정 엔드포인트 추가 완료!

Juser0 reviewed (commented):
수고했고

내가 중복된 부분에 대해서는 코멘트를 달진 않아서 코멘트 단 부분 참고해서 형이 봐줘!
지금 전체적으로 궁금한 내용은
- ResponseDto에 대한 생성로직을 계속 명시하는 식으로 구현한 이유는?
- RequestDto의 필드를 기반으로 Entity를 생성하는 로직을 계속 명시하는 식으로 구현한 이유는?
이 부분이야!

이 부분에 대한 답변과 코멘트부분 수정하구 Swagger에 설명까지 명시해서 진행하면 될 거 같아!

jang-namu committed:
[Refactor] Domain - Task: 불필요한 어노테이션 수정.
@AllArgsConstructor -> 생성자 명시적으로 선언.
클래스 수준의 @Builder 어노테이션에서 생성자 수준으로 변경.

'isCompleted' 필드 @setter -> updateIsCompleted()
행위를 명시적으로 나타낼 수 있는 메서드로 변경

jang-namu commented:
> 수고했고
> 
> 내가 중복된 부분에 대해서는 코멘트를 달진 않아서 코멘트 단 부분 참고해서 형이 봐줘! 지금 전체적으로 궁금한 내용은
> 
> * ResponseDto에 대한 생성로직을 계속 명시하는 식으로 구현한 이유는?
> * RequestDto의 필드를 기반으로 Entity를 생성하는 로직을 계속 명시하는 식으로 구현한 이유는?
>   이 부분이야!
> 
> 이 부분에 대한 답변과 코멘트부분 수정하구 Swagger에 설명까지 명시해서 진행하면 될 거 같아!

DTO <-> Entity 간 변환 방법은 Mapper 클래스, 사용자 정의, Mapper 라이브러리가 있다곤 하는데 아직 뭐가 좋은지 잘 모르겠어서
우선 명시적으로 작성했었어.

~~지금은 일단 Entity의 static 'toResponseDto' 메서드 | RequestDto에 static 'toEntity' 메서드를 정의해서 변경했어. 조만간 다시 올릴게!~~

2023.11.07) 전반적인 수정완료! 
Keep-Track: UserController - Update 엔드포인트 추가필요

jang-namu committed:
[Rename] TaskDto -> TaskRequestDto

jang-namu committed:
[Refactor] Task 도메인 관련 Entity <-> Dto 로직 메서드로 작성
Task-TaskResponseDto-TaskService 부분 수정.

jang-namu committed:
[Refactor] User 도메인 Entity <-> Dto 간 변환 로직, 메서드 작성

jang-namu committed:
[Feat] Task 전체 수정 엔드포인트 추가, 일부 엔드포인트 변경
'/tasks/iscompleted' -> '/tasks/completed'

jang-namu committed:
[Fix] UserResponseDto @Getter 어노테이션 추가
ObjectMapper가 Dto 생성 시 필드를 주입하지 못했던 문제 수정

jang-namu committed:
[Chore] POST '/users' 엔드포인트 반환 상태코드 변경 OK -> CREATED

jang-namu commented:
궁금한게 있습니다.

TaskResponseDto 클래스
<img width="882" alt="image" src="https://github.com/inu-appcenter/server-study-15.5/assets/121238128/016cc936-5043-44e1-883b-04c97ba89244">

Task 클래스
<img width="542" alt="image" src="https://github.com/inu-appcenter/server-study-15.5/assets/121238128/818e7d9f-1c0c-4b39-8595-629ab3f7d57b">

Task Entity <-> Dto 간 변화하는 로직을 메서드로 작성했는데
이 경우 User 엔티티의 기본키 userId를 가져와야 합니다.
Task와 User는 로딩 시점을 Lazy로 걸어놓고 연관관계 매핑을 했는데,
이렇게 되면 대부분의 엔드포인트는 쿼리가 두번 나가게되서 Lazy가 오히려 비효율적이 아닌가? 싶습니다.

jang-namu committed:
[Chore] swagger API 명세 및 Model 명세 추가

elyudwo commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/task/Task.java:
```diff
@@ -23,5 +31,47 @@ public class Task {
+        this.isCompleted = !this.isCompleted;
+    }
+
+    public static TaskResponseDto toResponseDto(Task task) {
+        return TaskResponseDto.builder()
+                .taskId(task.getTaskId())
+                .title(task.getTitle())
+                .description(task.getDescription())
+                .deadline(task.getDeadline())
+                .userId(task.getUser().getUserId())
+                .build();
+    }
```
정적 메소드로 만드신 이유가 궁금합니다!

jang-namu replied:
Java의 형변환과 같은 API를 참조해서 Intger.parseInt(), String.valueOf()와 같이 인스턴스보다 클래스 수준에서 책임을 갖는 것이 맞다고 생각했는데, 
지금와서 생각해보니 위와 같은 경우는 다른 클래스 -> 현재 클래스로의 변환이니 static이 자연스러운반면, 제 경우에는 현재 클래스의 데이터를 그대로 사용하니 static이 안 붙는게 자연스러울 수 있을 것 같네요. 
ModelMapper를 구현할 떄나 정적메서드로 써야겠군요. 감사합니다.

elyudwo commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/task/TaskService.java:
```diff
@@ -0,0 +1,80 @@
+    private final TaskRepository taskRepository;
+    private final UserRepository userRepository;
+
+    public List<TaskResponseDto> getAllByUserId(Long userId) throws Exception {
```
파라미터에 final을 붙이시는것도 고려해보시면 좋을 것 같습니다!

jang-namu replied:
전체적으로 파라미터에 변경은 고려되지 않으니 그러는게 좋겠네요. 감사합니다!

elyudwo commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/task/TaskService.java:
```diff
@@ -0,0 +1,80 @@
+            throw new Exception("User를 찾을 수 없습니다.");
+        }
+
+        taskResponseDtos = taskList.stream().map(TaskResponseDto::new).collect(Collectors.toList());
```
혹시 이렇게 넘겨주면 unmodifiable한 상태로넘겨주게 되는것인가요??

jang-namu replied:
TaskResponseDto에는 setter나 modify 역할을 하는 메서드가 없어 변경이 불가할 것이라고 생각됩니다!

jang-namu replied:
이해를 제대로 못 했었군요. taskResonseDtos를 그대로 반환하기 때문에 
Collections.unmodifiableCollection()의 효과는 없습니다.
즉, 이를 컨트롤러에서 받아 수정이 가능합니다.

elyudwo commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/task/TaskService.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    public TaskResponseDto save(TaskRequestDto taskRequestDto) throws Exception {
+        Optional<User> selectedUser = userRepository.findById(taskRequestDto.getUserId());
+        selectedUser.orElseThrow(() -> new Exception("해당하는 사용자를 찾지 못 함."));
```
아래 코드처럼 한번에 던지는 방법도 좋을 것 같습니다!
`User selectedUser = userRepository.findById(taskRequestDto.getUserId()).orElseThrow()`

jang-namu replied:
가독성을 고려해서 평소 나눠서 작성하는 것을 선호했는데, 
오히려 이상할까요? 가독성이 떨어진다면 수정하겠습니다!
의견부탁드립니다 :)

jang-namu replied:
생각해보면 참조가 유지되는 동안 더 이상 사용되지 않을 Optional 객체가 존재하게 되는군요.

elyudwo replied:
제 개인적인 생각으로는 가독성이 떨어진다고 생각하는데 이건 어디까지나 지극히 개인적인 생각이라 철희님 주관대로 하시는게 좋을 것 같습니다 @

elyudwo commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/task/TaskService.java:
```diff
@@ -0,0 +1,80 @@
+    public TaskResponseDto updateTask(TaskRequestDto taskRequestDto) throws Exception {
+        Optional<Task> selectedTask = taskRepository.findById(taskRequestDto.getTaskId());
+
+        Task savedTask;
+        Task puttedTask;
```
메소드 내 변수의 참조를 한번에 할당하지 않으시고 따로 나누시는 이유가 궁금합니다!

jang-namu replied:
이 부분은 try-catch로 인해 savedTask, puttedTask가 초기화되지 않을 가능성이 있어서 분리했습니다. (컴파일 에러)

elyudwo commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/task/TaskService.java:
```diff
@@ -0,0 +1,80 @@
+
+        Task savedTask;
+        Task puttedTask;
+        try {
+            selectedTask.orElseThrow(() -> new Exception("해당하는 Task 정보를 찾을 수 없음"));
+            puttedTask = selectedTask.get();
+            puttedTask.updateFromDto(taskRequestDto);
+            savedTask = taskRepository.save(puttedTask);
+        } catch (Exception taskNotFoundException) {
+            Optional<User> taskOwner = userRepository.findById(taskRequestDto.getUserId());
+            taskOwner.orElseThrow(() -> new Exception("해당 Task와 User의 정보를 찾을 수 없음"));
+            puttedTask = TaskRequestDto.toEntity(taskRequestDto, taskOwner.get());
+            savedTask = taskRepository.save(puttedTask);
+        }
+        return Task.toResponseDto(savedTask);
+    }
```
try catch로 잡는 것 보다 if 문을 이용해 메소드를 두개로 나누는건 어떠신가요??

jang-namu replied:
3개의 방법으로 작성해보았는데, 혹시 어떤 방식이 읽으실 때 편하신가요?

1차  

![image](https://github.com/inu-appcenter/server-study-15.5/assets/121238128/25b02ae0-724c-4c1d-8e9a-cd7227817e3e)

2차  

![image](https://github.com/inu-appcenter/server-study-15.5/assets/121238128/d7715d5c-393c-48dd-a97c-155114192ef2)

3차  

<img width="911" alt="스크린샷 2023-11-14 오전 1 28 57" src="https://github.com/inu-appcenter/server-study-15.5/assets/121238128/bcc5475a-57d6-40e6-95df-cdacf85b6102">

elyudwo commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/task/dto/TaskResponseDto.java:
```diff
@@ -0,0 +1,49 @@
+        this.deadline = deadline;
+    }
+
+    public TaskResponseDto(Task task) {
+        this.taskId = task.getTaskId();
+
+        // 이렇게되면 ReponseDTO에 필요없는 User 정보가 포함되지는 않나?
+        // 로딩시점을 Lazy로 설정해놓고도, DTO를 만들기 위해 User 엔티티를 조회하게 되는거 아닌가.
+        this.userId = task.getUser().getUserId();
+        this.title = task.getTitle();
+        this.description = task.getDescription();
+        this.deadline = task.getDeadline();
+    }
```
클라이언트 측에서 이미 userId를 알고있는 상태에서 request를 보내는데 ResponseDto 에서 다시 userId를 반환해주는 이유가 궁금합니다!

jang-namu replied:
생각해보니, 이것도 TaskController에 요청이 들어오는 문맥상 불필요한 정보인 것 같네요. userId를 들어내겠습니다.

elyudwo reviewed (commented):
설계가 잘 돼있어 코드 읽기가 편했습니다 잘 읽었습니다!
